src/lib.rs

```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+    pub pre_identity_object: PreIdentityObject,
+    #[serde(rename = "attributeList")]
+    pub attribute_list: AttributeList,
+    #[serde(rename = "signature")]
+    pub signature_hex: String,
+}
+
+#[derive(Debug, Serialize)]
+pub struct PreIdentityObject {
+    #[serde(rename = "idCredPub")]
+    pub id_cred_pub_hex: String,
+    #[serde(rename = "ipArData")]
+    pub ip_ar_data: HashMap<ArIdentity, ArData>,
+    #[serde(rename = "choiceArData")]
+    pub choice_ar_data: ChoiceArParameters,
+    #[serde(rename = "idCredSecCommitment")]
+    pub id_cred_sec_commitment_hex: String,
+    #[serde(rename = "prfKeyCommitmentWithIP")]
+    pub prf_key_commitment_with_ip_hex: String,
+    #[serde(rename = "prfKeySharingCoeffCommitments")]
+    pub prf_key_sharing_coeff_commitments_hex: Vec<String>,
+    #[serde(rename = "proofsOfKnowledge")]
+    pub proofs_of_knowledge_hex: String,
+}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub struct ChoiceArParameters {
+    #[serde(rename = "arIdentities")]
+    pub ar_identities: Vec<ArIdentity>,
+    #[serde(rename = "threshold")]
+    pub threshold: u32,
+}
+
+#[derive(Debug, Serialize)]
+pub struct ArData {
+    #[serde(rename = "encPrfKeyShare")]
+    pub enc_prf_key_share_hex: String,
+    #[serde(rename = "proofComEncEq")]
+    pub proof_com_enc_eq_hex: String,
+}
+
+#[derive(Debug, Serialize)]
+pub struct AttributeList {
+    #[serde(rename = "validTo")]
+    pub valid_to_year_month: String,
+    #[serde(rename = "createdAt")]
+    pub created_at_year_month: String,
+    #[serde(rename = "maxAccounts")]
+    pub max_accounts: u8,
+    #[serde(rename = "chosenAttributes")]
+    pub chosen_attributes: HashMap<String, String>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct CredentialPublicKeys<Key> {
+    #[serde(rename = "keys")]
+    pub keys: HashMap<KeyIndex, Key>,
+    #[serde(rename = "threshold")]
+    pub threshold: u8,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct VerifyKeyWithScheme {
+    #[serde(rename = "schemeId")]
+    pub scheme_id: String,
+    #[serde(rename = "verifyKey")]
+    pub key_hex: String,
+}
+
+pub type CredentialPublicKeysHex = CredentialPublicKeys<String>;
+pub type CredentialPublicKeysWithScheme = CredentialPublicKeys<VerifyKeyWithScheme>;
+
+/* OUTPUTS */
+
+#[derive(Debug, Deserialize)]
+pub struct AccountCredentialResult {
+    #[serde(rename = "randomness")]
+    randomness: Randomness,
+    #[serde(rename = "unsignedCdi")]
+    credential: AccountCredential,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Randomness {
+    #[serde(rename = "attributesRand")]
+    pub attributes_rand_hex: HashMap<String, String>,
+    #[serde(rename = "credCounterRand")]
+    pub cred_counter_rand_hex: String,
+    #[serde(rename = "idCredSecRand")]
+    pub id_cred_sec_rand_hex: String,
+    #[serde(rename = "maxAccountsRand")]
+    pub max_accounts_rand_hex: String,
+    #[serde(rename = "prfRand")]
+    pub prf_rand_hex: String,
+}
```

Comment:
This also seems to be duplicating things by replacing types with Strings.

Reply:
Yes, in order to conform with UniFFI requirements. Remember that these types exist for being used in the other side of FFI. There's nothing other than string that reasonably makes sense across that boundary.

Reply:
Right, but now you've duplicated everything, documetned nothing, and manually writtent this both in Rust and in the interface description file.

There is also no clearly stated relationship between any of these types and the ones defined in base.

If this UniFFI can generate types, why are we manually defining them here?

Also, the way we convert from imprecise types to precise ones in other places, e.g., when using grpc generated types, is to have TryFrom or From instances that establish a precise relationship between types. It also means that we get compilation errors if either the generated ones or  the more precise defined ones change.

So why can't we do that here? As I read this https://mozilla.github.io/uniffi-rs/tutorial/Rust_scaffolding.html that should be the intended way to do it. Or what am I missing?